### PR TITLE
add type annotations to torch.nn.modules.pooling

### LINKED ISSUE #49503: Enable torch.nn.modules.pooling typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.nn.modules.pooling]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
closes gh-49503

### CODE REVIEW

**Review by walterddr — APPROVED** [review]
lgtm. 
nit: seems like we can stripe the definition in `AvgPool1d` as well (line 519-521)

File: torch/nn/modules/pooling.py [hunk 1]
```diff
@@ -739,10 +739,10 @@ class FractionalMaxPool2d(Module):
     __constants__ = ['kernel_size', 'return_indices', 'output_size',
                      'output_ratio']
 
-    kernel_size: _size_2_t
+    kernel_size: Tuple[int, int]
     return_indices: bool
-    output_size: _size_2_t
-    output_ratio: _ratio_2_t
+    output_size: Optional[Tuple[int, int]]
+    output_ratio: Optional[Tuple[int, int]]
```
Comment (rgommers) [inline torch/nn/modules/pooling.py:745]: ratio is tuple of `float` not `int`

File: torch/nn/modules/pooling.py [hunk 2]
```diff
@@ -802,10 +802,10 @@ class FractionalMaxPool3d(Module):
     """
     __constants__ = ['kernel_size', 'return_indices', 'output_size',
                      'output_ratio']
-    kernel_size: _size_3_t
+    kernel_size: Tuple[int, int, int] 
     return_indices: bool
-    output_size: _size_3_t
-    output_ratio: _ratio_3_t
+    output_size: Optional[Tuple[int, int, int]]
+    output_ratio: Optional[Tuple[int, int, int]]
```
Comment (rgommers) [inline torch/nn/modules/pooling.py:808]: same here, `int` -> `float`

**Review by rgommers — CHANGES_REQUESTED** [review]
Need a small tweak, ratios are floats.

File: torch/nn/modules/pooling.py (hunk 1 above)
Comment (walterddr) [reply torch/nn/modules/pooling.py:745]: good catch!

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
@walterddr would you be able to land this PR?

**Comment by guilhermeleobas** [comment]
Do not merge this PR until one checks if the annotations introduce any regression. See:
https://github.com/pytorch/pytorch/pull/49564#issuecomment-752916072

File: test/test_jit.py [hunk 3]
```diff
@@ -926,6 +926,53 @@ def forward(self, input):
...
+        m2 = Mod(nn.MaxPool2d(2, stride=2, return_indices=True),
+                 nn.MaxUnpool2d(2, stride=2))
+        inp2d = torch.arange(1, 17, dtype=torch.float).reshape(1, 1, 4, 4)
+        self.checkModule(m2, (inp2d,))
+
+        # crashes pytorch
+        # m3 = Mod(nn.MaxPool3d(3, stride=2, return_indices=True), 
+        #          nn.MaxUnpool3d(3, stride=2))
+        # inp3d = torch.randn(20, 16, 51, 33, 15)
+        # self.checkModule(m3, (inp3d,))
```
Comment (guilhermeleobas) [inline test/test_jit.py:955]: This test crashes the PyTorch jit:
```
terminate called after throwing an instance of 'c10::Error'
  what():  found an invalid max index 24225 (output volumes are of size 50x32x14
Exception raised from max_unpooling3d_forward_out_cpu_frame at ../aten/src/ATen/native/MaxUnpooling.cpp:196 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x68 (0x7fc38ea7da08 in /home/guilhermel/git/pytorch/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xfa (0x7fc38ea42804 in /home/guilhermel/git/pytorch/torch/lib/libc10.so)
frame #2: <unknown function> + 0x14d1fcd (0x7fc39c670fcd in /home/guilhermel/git/pytorch/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x86438 (0x7fc3af2b4438 in /home/guilhermel/.conda/envs/pytorch-cuda-dev/lib/libgomp.so.1)
frame #4: __kmp_invoke_microtask + 0x93 (0x7fc3af2c8c83 in /home/guilhermel/.conda/envs/pytorch-cuda-dev/lib/libgomp.so.1)
frame #5: <unknown function> + 0x3c747 (0x7fc3af26a747 in /home/guilhermel/.conda/envs/pytorch-cuda-dev/lib/libgomp.so.1)
frame #6: <unknown function> + 0x3b758 (0x7fc3af269758 in /home/guilhermel/.conda/envs/pytorch-cuda-dev/lib/libgomp.so.1)
frame #7: <unknown function> + 0x855aa (0x7fc3af2b35aa in /home/guilhermel/.conda/envs/pytorch-cuda-dev/lib/libgomp.so.1)
frame #8: <unknown function> + 0x9609 (0x7fc3cbc5e609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #9: clone + 0x43 (0x7fc3cbb85293 in /lib/x86_64-linux-gnu/libc.so.6)

Aborted (core dumped)
```

**Comment by guilhermeleobas** [comment]
I've fixed a few type annotations that were crashing the jit and added tests

**Comment by guilhermeleobas** [comment]
I guess part of this PR is blocked by https://github.com/pytorch/pytorch/pull/47888

Issue https://github.com/pytorch/pytorch/issues/45904 summarizes the error that I'm facing at the moment to add tests for `nn.MaxUnpoolNd`:

```python
class Mod(nn.Module):
    def __init__(self\):
        super().__init__()
        self.pool = nn.MaxPool1d(2, stride=2, return_indices=True)
        self.unpool = nn.MaxUnpool1d(2, stride=2)

    def forward(self, input):
        output, indices = self.pool(input)
        return self.unpool(output, indices)
```

When `return_indices=True`, `nn.MaxPool1d.forward` will return a tuple of tensors rather than a single tensor (current annotation).

cc @rgommers

**Comment by rgommers** [comment]
> When `return_indices=True`, `nn.MaxPool1d.forward` will return a tuple of tensors rather than a single tensor (current annotation).

As noted in https://github.com/pytorch/pytorch/issues/45904#issuecomment-704968488, that requires adding a `Literal[True]` overload. 

Best to wait with updating this PR until gh-47888 is landed I think.

**Pull request closed without merging**